### PR TITLE
APP-157: [C] Basic level Apps: Data collection

### PR DESCRIPTION
**Pull request by RostakaGmfun**
/cc @Slash32 @Dubland @vadimol

### CODE REVIEW

File: datacollectiondemo/source/c/src/kaa_demo.c [hunk 1]
```diff
@@ -14,185 +14,163 @@
...
- * Forward declarations.
- */
-extern kaa_error_t ext_limited_log_storage_create(void **log_storage_context_p
-                          , kaa_logger_t *logger
-                          , size_t size
-                          , size_t percent);
+#define DEMO_LOG_UPLOAD_THRESHOLD 5
 
+typedef struct
+{
```
Comment (rasendubi) [inline datacollectiondemo/source/c/src/kaa_demo.c]: code style. should be:

``` c
typedef struct {
```

File: datacollectiondemo/source/c/src/kaa_demo.c [hunk 2]
```diff
@@ -14,185 +14,163 @@
...
+    kaa_client_t *kaa_client;
+    int32_t sample_period;
+    time_t last_upload_time;
+} temperature_sensor_context;
 
-static kaa_client_t *kaa_client = NULL;
+static void error_cleanup(kaa_client_t *client)
+{
+    if (client) {
```
Comment (rasendubi) [inline datacollectiondemo/source/c/src/kaa_demo.c]: Compare with null explicitly

File: datacollectiondemo/source/c/src/kaa_demo.c [hunk 3]
```diff
@@ -14,185 +14,163 @@
...
+    if (current - sensor_context->last_upload_time >=
+            sensor_context->sample_period) {
+        send_temperature(sensor_context->kaa_client);
+        sensor_context->last_upload_time = time(NULL);
     }
+}
 
-    log_record->level = ENUM_LEVEL_KAA_INFO;
-    log_record->tag = kaa_string_move_create(KAA_DEMO_LOG_TAG, NULL);
+int main()
```
Comment (rasendubi) [inline datacollectiondemo/source/c/src/kaa_demo.c]:
``` c
int main(void)
```

File: datacollectiondemo/source/c/src/kaa_demo.c [hunk 4]
```diff
@@ -14,185 +14,163 @@
...
+{
+    printf("Data collection demo started\n");
+
+    kaa_client_t *kaa_client = NULL;
+    temperature_sensor_context sensor_context;
+    kaa_error_t error;
+    void *log_upload_strategy_context = NULL;
 
-    log_record->time_stamp = KAA_TIME() * 1000;
+    error = kaa_client_create(&kaa_client, NULL);
```
Comment (rasendubi) [inline datacollectiondemo/source/c/src/kaa_demo.c]: Use declaration with initialization

File: datacollectiondemo/source/c/src/kaa_demo.c [hunk 5]
```diff
@@ -14,185 +14,163 @@
...
     }
+}
 
-    log_record->level = ENUM_LEVEL_KAA_INFO;
-    log_record->tag = kaa_string_move_create(KAA_DEMO_LOG_TAG, NULL);
+int main()
+{
+    printf("Data collection demo started\n");
+
+    kaa_client_t *kaa_client = NULL;
```
Comment (rasendubi) [inline datacollectiondemo/source/c/src/kaa_demo.c:107]: Move declaration closer to the place it's initialized

File: datacollectiondemo/source/c/src/kaa_demo.c [hunk 6]
```diff
@@ -14,185 +14,163 @@
...
-    char log_message_buffer[KAA_DEMO_LOG_BUF_SZ];
-    snprintf(log_message_buffer, KAA_DEMO_LOG_BUF_SZ, KAA_DEMO_LOG_MESSAGE"%zu", log_record_counter);
+    if (error) {
+        fprintf(stderr, "Failed to create Kaa client\n");
+        return EXIT_FAILURE;
+    }
+
+    kaa_configuration_root_receiver_t receiver = {
+        &sensor_context,
+        configuration_update
```
Comment (rasendubi) [inline datacollectiondemo/source/c/src/kaa_demo.c]: add trailing comma

**Comment by Slash32** [comment]
LGTM

**Pull request merged**